src/modules/Elsa.Workflows.Runtime/Services/DefaultWorkflowRuntime.cs

```diff
@@ -90,6 +103,70 @@ public async Task<WorkflowExecutionResult> StartWorkflowAsync(string definitionI
     {
         return await StartWorkflowAsync(definitionId, options);
     }
+    
+    /// <inheritdoc />
+    public async Task CancelWorkflowAsync(string workflowInstanceId, CancellationToken cancellationToken)
+    {
+        //Should we add a cancellationtoken?
+        var workflowExecutionContext = await _workflowExecutionContextStore.FindAsync(workflowInstanceId);
+        
+        if (workflowExecutionContext is null)
+        {
+            // The execution context is not running on this instance.
+            // It might not be running on any instance, so check the db and update the record.
+            // Use lock to prevent race conditions and other instances from updating the workflow context
+            await using var cancelLock = await _distributedLockProvider.TryAcquireLockAsync($"{workflowInstanceId}-cancel", cancellationToken: cancellationToken);
+            if (cancelLock == null)
+                return;
+            
+            var workflowInstance = await _workflowInstanceStore.FindAsync(workflowInstanceId, cancellationToken);
+            if (workflowInstance is null
+                || workflowInstance.SubStatus == WorkflowSubStatus.Cancelled
+                || workflowInstance.SubStatus == WorkflowSubStatus.Faulted)
+                return;
+
+            var workflowState = await ExportWorkflowStateAsync(workflowInstanceId, cancellationToken);
+
+            if (workflowState == null)
+                throw new Exception("Workflow state not found");
+
+            var workflowDefinition = await _workflowDefinitionService.FindAsync(workflowState.DefinitionId, VersionOptions.SpecificVersion(workflowState.DefinitionVersion), cancellationToken);
+
+            if (workflowDefinition == null)
+                throw new Exception("Workflow definition not found");
+
+            var workflow = await _workflowDefinitionService.MaterializeWorkflowAsync(workflowDefinition, cancellationToken);
+            workflowExecutionContext = await WorkflowExecutionContext.CreateAsync(_serviceProvider, workflow, workflowState, cancellationTokens: cancellationToken);
+
+            //Is this still necessary?
```

Comment:
Good point. It looks like this is not necessary, given that the `await WorkflowExecutionContext.CreateAsync` call on line 139 already maps the state into the context.

Reply:
Removed the Apply call